src/System.Net.Security/tests/FunctionalTests/NegotiateStreamInvalidOperationTest.cs

```diff
@@ -0,0 +1,274 @@
+                Assert.Throws<NotImplementedException>(() => client.Position);
+                Assert.Throws<NotSupportedException>(() => client.Seek(0, new SeekOrigin()));
+
+                Task[] auth = new Task[2];
+                auth[0] = client.AuthenticateAsClientAsync();
+                auth[1] = server.AuthenticateAsServerAsync();
+
+                await TestConfiguration.WhenAllOrAnyFailedWithTimeout(auth);
+
+                Assert.True(client.CanRead);
+                Assert.True(client.CanWrite);
+                Assert.True(server.CanRead);
+                Assert.True(server.CanWrite);
+            }
+        }
+
+        [Fact]
+        public async Task NegotiateStream_EndReadEndWriteInvalidParameter_Throws()
+        {
+            byte[] recvBuf = new byte[_sampleMsg.Length];
+            VirtualNetwork network = new VirtualNetwork();
+
+            using (var clientStream = new VirtualNetworkStream(network, isServer: false))
+            using (var serverStream = new VirtualNetworkStream(network, isServer: true))
+            using (var client = new NegotiateStream(clientStream))
+            using (var server = new NegotiateStream(serverStream))
+            {
+                Task[] auth = new Task[2];
+                auth[0] = client.AuthenticateAsClientAsync(CredentialCache.DefaultNetworkCredentials, string.Empty);
+                auth[1] = server.AuthenticateAsServerAsync();
+                await TestConfiguration.WhenAllOrAnyFailedWithTimeout(auth);
+
+                auth[0] = Task.Factory.FromAsync(client.BeginWrite, CustomEndWrite, _sampleMsg, 0, _sampleMsg.Length, client);
+                auth[1] = Task.Factory.FromAsync(server.BeginRead, CustomEndRead, recvBuf, 0, _sampleMsg.Length, server);
+                await TestConfiguration.WhenAllOrAnyFailedWithTimeout(auth);
+            }
+        }
+
+        [Fact]
+        public async Task NegotiateStream_ConcurrentAsyncReadOrWrite_ThrowsNotSupportedException()
+        {
+            byte[] recvBuf = new byte[_sampleMsg.Length];
+            VirtualNetwork network = new VirtualNetwork();
+
+            using (var clientStream = new VirtualNetworkStream(network, isServer: false))
+            using (var serverStream = new VirtualNetworkStream(network, isServer: true))
+            using (var client = new NegotiateStream(clientStream))
+            using (var server = new NegotiateStream(serverStream))
+            {
+                Task[] auth = new Task[2];
+                auth[0] = client.AuthenticateAsClientAsync(CredentialCache.DefaultNetworkCredentials, string.Empty);
+                auth[1] = server.AuthenticateAsServerAsync();
+                await TestConfiguration.WhenAllOrAnyFailedWithTimeout(auth);
+
+                // CustomEndWrite/Read will not reset the variable which monitors concurrent write/read.
+                auth[0] = Task.Factory.FromAsync(client.BeginWrite, CustomEndWrite, _sampleMsg, 0, _sampleMsg.Length, client);
```

Comment:
CustomEndWrite feels odd here.  It's doing its own unrelated testing that happens to avoid calling EndWrite on the client, and that's then being used here to test something else.  I suggest you split that testing off to be separate, and here just use a lambda that doesn't call client.EndWrite.